### PR TITLE
fix: Negative collection_offset Causes Multiple Executions

### LINKED ISSUE #18175: Negative collection_offset Causes Multiple Executions
### Relevant telegraf.conf

```toml
/etc/telegraf# cat telegraf.conf
[[inputs.exec]]
  interval = "30s"
  commands = ["python3 /etc/telegraf/test.py"]
  timeout = "3s"
  collection_offset = "-15s"
  data_format = "influx"

[[outputs.file]]
  files = ["stdout", "/etc/telegraf/test_output.txt"]


/etc/telegraf# cat test.py 
print("metric,tag1=1,tag2=2 value=1")
```

### Logs from Telegraf

```text
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
2025-12-31T13:07:48Z D! [inputs.exec] Previous collection has not completed; scheduled collection skipped
```

### System info

Ubuntu 24.04.3 , Telegraf 1.37.0

### Docker

When running with a negative offset Telegraf is executing the python script multiple times, when running with a positive offset there are no issues.

### Steps to reproduce

1. Copy telegraf.conf config above to /etc/telegraf folder.
2. Copy python script to /etc/telegraf folder and make executable.
3. Run /usr/bin/telegraf --config telegraf.conf with the config of collection_offset = "-15s" to see the problem, run the same command again with collection_offset = "5s" and the issue no longer happens.
...

### Expected behavior

I would expect telegraf to run the command / script once at 15s before the set interval rate.

### Actual behavior

I appears to be running it multiple times and printing the results multiple times.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by majiayu000**
Fixes #18175

## Changes
- Add check in AlignedTicker.next() to ensure duration is positive after applying negative offset
- Update sleep() to handle negative durations by returning immediately
- Add tests for negative offset behavior

### CODE REVIEW

**Comment by Hipska** [comment]
Shouldn't the solution be to let Telegraf error out when a negative duration is configured anywhere?

**Comment by srebhan** [comment]
@majiayu000 adding to the comment of @Hipska, what is the reasoning for a negative offset?

I'm asking because the effect is basically the same as doing `interval - abs(negative_offset)` isn't it? So maybe that's an easier and more elegant solution?

**Comment by majiayu000** [comment]
@Hipska @srebhan Thanks! I updated the logic to normalize negative offsets into [0, interval), which is equivalent to `interval - abs(offset)` for negative values. This makes the behaviorexplicit and avoids the rapid-fire tick issue without changing scheduling semantics.

File: agent/tick.go [hunk 1]
```diff
@@ -78,6 +78,13 @@ func (t *AlignedTicker) next(now time.Time) time.Duration {
 		d = t.interval
 	}
 	d += t.offset
+
+	// Ensure the duration is positive before adding jitter. This can still
+	// happen with small intervals and clock adjustments.
+	if d <= 0 {
+		d += t.interval
+	}
```
Comment (srebhan) [inline agent/tick.go:86]: This can never happen, can it?

File: agent/tick.go [hunk 2]
```diff
@@ -132,12 +139,23 @@ func NewUnalignedTicker(interval, jitter, offset time.Duration) *UnalignedTicker
...
+		offset:   normalizeOffset(offset, interval),
 	}
 	t.start(clock.New())
 	return t
 }
 
+func normalizeOffset(offset, interval time.Duration) time.Duration {
+	if interval <= 0 || offset == 0 {
+		return offset
+	}
```
Comment (srebhan) [inline agent/tick.go:151]: Please error out during configuration if the interval is negative! This is clearly a misconfiguration!

File: agent/tick.go [hunk 3]
```diff
@@ -132,12 +139,23 @@ func NewUnalignedTicker(interval, jitter, offset time.Duration) *UnalignedTicker
...
+	offset = offset % interval
```
Comment (srebhan) [inline agent/tick.go:152]: Isn't this unexpected to the user? I rather prefer to error out during configuration if the offset is larger than the interval...

File: agent/tick.go [hunk 4]
```diff
@@ -132,12 +139,23 @@ func NewUnalignedTicker(interval, jitter, offset time.Duration) *UnalignedTicker
...
+	if offset < 0 {
+		offset += interval
+	}
```
Comment (srebhan) [inline agent/tick.go:155]: Please make sure that the result is positive during configuration and error out otherwise!

**Comment by majiayu000** [comment]
@srebhan Thanks for the detailed review feedback! I've updated the PR to address all your points:

**Changes made:**
1. **Negative offsets are now rejected entirely** with a clear error message suggesting `interval - offset` as the alternative (e.g., `negative collection_offset is not supported, use collection_offset = "7s" instead of "-3s"`)
2. **Offset exceeding interval is rejected** with error: `collection_offset 15s exceeds interval 10s`
3. **Negative interval is rejected** with error: `interval must be a positive duration`
4. Removed the previous `normalizeOffset` approach that wrapped negative values
5. Updated docs to state offsets must be non-negative and not exceed the interval
6. Added comprehensive validation tests covering all error cases

The validation happens at ticker creation time, so misconfiguration is caught early and logged as an error.

Could you please reopen this PR? Thank you!
